app/controllers/concerns/phone_verification.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+module PhoneVerification
```

Comment:
Bu concern şu şekliyle iki farklı süreci yönetmeye çalışıyor. Bu süreçler telefon doğrulama bakımından sadece birbirleriyle yakınlık gösteriyor diğer türlü birbirinden tamamıyla farklı işler yapıyorlar.

Bu bakımdan bu concern'de sadece telefon doğrulama için gerekli helper metodları tutmamız gerekiyor. Belki hiç bu modulü kullanmadan telefon doğrulamayı her bir controller için ayrı ayrı el almalıyız.

Dikkat edersen, controller kontrolü yaparak farklı işleri yönetmeye çalışıyoruz. Bu kod okunurluğunu zorlaştırıyor.

Reply:
Örnek olması amacıyla;

**concerns/phone_verification.rb**
```ruby

module PhoneVerification
  extend ActiveSupport::Concern

  private

  def send_verification_code(phone)
    Twilio::Verify.send_phone_verification_code(phone) == 'ok'
  end

  def phone_verified?
    Twilio::Verify.check_verification_code(
      params_for_phone_verification[:mobile_phone],
      params_for_phone_verification[:verification_code]
    )
  end

  def params_for_phone_verification
    params.require(:phone_verification).permit(:mobile_phone, :verification_code)
  end
end
```

**registrations_controller**
```ruby
def phone_verification
  current_user.assign_attributes(params_for_phone_update)

  if current_user.valid? && current_user.mobile_phone_changed?
    flash[:alert] = t('errors.system_error') unless send_verification_code(params_for_phone_update[:mobile_phone])
  end

  respond_to :js
end

def update_phone
  response = phone_verified?
  unless response == 'ok'
    return redirect_to account_path,
                       alert: t("twilio.errors.#{response.error_code}", default: t('errors.system_error'))
  end

  if current_user.update(mobile_phone: params_for_phone_verification[:mobile_phone])
    flash[:notice] = t('.success')
  else
    flash[:alert] = t('errors.system_error')
  end
  redirect_to account_path
end

private

def params_for_phone_update
  params.require(:user).permit(:mobile_phone)
end
```

Reply:
Bu ilk gerçeklememin mantığıyla aynı(ilk WIP komiti ile atmıştım). Burada ise aynı işlemin yapıldığı yerleri concern'de toplamaya çalıştım. Süreç olarak yapılan işlerde aslında hiç bir fark yok sadece update edilen veriler farklılık gösteriyor. Ancak controller kontrolü pek uygun olmadı farkındayım. Bunları göz önünde bulundurarak bir düzenleme yapacağım.

Reply:
Gitlab projesinde concern içinde controller kontrolü yapılmış: https://gitlab.com/gitlab-org/gitlab-ce/blob/master/app/controllers/concerns/preview_markdown.rb#L11 Hatta tek bir controller'a özel fonksiyon da tanımlanmış. Ancak uygun bir pratik mi bilemiyorum :thinking:

Reply:
Bu concern isim olarak `PhoneVerification` bu bağlamdan bakınca sadece yapması gereken iş, telefon doğrulama ile ilgili lojiği içerinde barındırmak gibi geliyor bana, fakat koda baktığımızda aktivasyon ile bir kısım lojiğinde burada olduğunu görüyorum.

Reply:
bence bu karışıklık concern olmasından kaynaklı. görünen o ki bu iki controller dışında yeni bir controller buraya eklenmeyecek. sadece iki controller'daki ortaklıklar (aslında @isubas ın dediği için farklılıklar da var) için concern kullanmak, bu ortaklıkları yönetme maliyeti ortaya çıkarmış gibi görünüyor